File: src/main/java/cn/guajichi/idlepool/util/DurationParser.java
package cn.guajichi.idlepool.util;

import java.time.Duration;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class DurationParser {
    private static final Pattern PART = Pattern.compile("(\\d+)([dhms])");

    private DurationParser() {
    }

    public static Duration parse(String input) {
        if (input == null || input.isBlank()) {
            throw new IllegalArgumentException("Duration cannot be blank");
        }

        String normalized = input.toLowerCase(Locale.ROOT).replace(" ", "");
        Matcher matcher = PART.matcher(normalized);
        long seconds = 0;
        int cursor = 0;
        while (matcher.find()) {
            if (matcher.start() != cursor) {
                throw new IllegalArgumentException("Invalid duration: " + input);
            }
            long value = Long.parseLong(matcher.group(1));
            seconds = Math.addExact(seconds, Math.multiplyExact(value, unitSeconds(matcher.group(2).charAt(0))));
            cursor = matcher.end();
        }
        if (cursor != normalized.length() || seconds <= 0) {
            throw new IllegalArgumentException("Invalid duration: " + input);
        }
        return Duration.ofSeconds(seconds);
    }

    private static long unitSeconds(char unit) {
        return switch (unit) {
            case 'd' -> 86_400;
            case 'h' -> 3_600;
            case 'm' -> 60;
            case 's' -> 1;
            default -> throw new IllegalArgumentException("Unsupported duration unit: " + unit);
        };
    }
}

File: src/main/kotlin/top/cnuo/idlepool/util/DurationParser.kt
package top.cnuo.idlepool.util

import java.time.Duration
import java.util.Locale

object DurationParser {
    private val part = Regex("(\\d+)([dhms])")

    @JvmStatic
    fun parse(input: String?): Duration {
        require(!input.isNullOrBlank()) { "Duration cannot be blank" }
        val normalized = input.lowercase(Locale.ROOT).replace(" ", "")
        var seconds = 0L
        var cursor = 0
        for (match in part.findAll(normalized)) {
            require(match.range.first == cursor) { "Invalid duration: $input" }
            val value = match.groupValues[1].toLong()
            seconds = Math.addExact(seconds, Math.multiplyExact(value, unitSeconds(match.groupValues[2][0])))
            cursor = match.range.last + 1
        }
        require(cursor == normalized.length && seconds > 0) { "Invalid duration: $input" }
        return Duration.ofSeconds(seconds)
    }

    private fun unitSeconds(unit: Char) = when (unit) {
        'd' -> 86_400L
        'h' -> 3_600L
        'm' -> 60L
        's' -> 1L
        else -> throw IllegalArgumentException("Unsupported duration unit: $unit")
    }
}
